### DISCUSSION TITLE
min() together with optional()

### DISCUSSION THREAD

**Discussion by medihack** [colinhacks/zod#553] [Q&A] [answered]
Is there a way to make a field optional, but if it is set, then it must be at least 8 chars long?

I tried `z.string().min(8).optional()`, but it requires the field to be set. Also tried `z.string().min(8).or(z.string().optional())`, but this allows shorter strings (e.g. 5 chars long).

Edit:
I decided to create an [issue](https://github.com/colinhacks/zod/issues/554) as the `optional` in the example above is simply ignored, but it seems to be correctly typed.

**Comment by scotttrinh** [comment]
Does `z.string().min(8).or(z.undefined())` work for your use case?

```ts
import { z } from "./index";

const maybeShortStr = z.string().min(5).or(z.undefined());

console.log(maybeShortStr.safeParse(undefined));
console.log(maybeShortStr.safeParse("1234"));
console.log(maybeShortStr.safeParse("12345"));
```

```
{ success: true, data: undefined }
{
  success: false,
  error: ZodError: [
    {
      "code": "too_small",
      "minimum": 5,
      "type": "string",
      "inclusive": true,
      "message": "Should be at least 5 characters",
      "path": []
    }
  ]
      at new ZodError (/Users/strinh/projects/zod/src/ZodError.ts:134:5)
      at handleResult (/Users/strinh/projects/zod/src/types.ts:76:19)
      at ZodUnion.ZodType.safeParse (/Users/strinh/projects/zod/src/types.ts:126:12)
      at Object.<anonymous> (/Users/strinh/projects/zod/src/playground.ts:6:27)
      at Module._compile (internal/modules/cjs/loader.js:1063:30)
      at Module.m._compile (/Users/strinh/projects/zod/node_modules/ts-node/src/index.ts:1056:23)
      at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
      at Object.require.extensions.<computed> [as .ts] (/Users/strinh/projects/zod/node_modules/ts-node/src/index.ts:1059:12)
      at Module.load (internal/modules/cjs/loader.js:928:32)
      at Function.Module._load (internal/modules/cjs/loader.js:769:14) {
    issues: [ [Object] ],
    format: [Function (anonymous)],
    addIssue: [Function (anonymous)],
    addIssues: [Function (anonymous)],
    flatten: [Function (anonymous)]
  }
}
{ success: true, data: '12345' }
```

**Reply by medihack** [reply]
Not really (at least not with changing some more stuff). The client (after validation with a react hook form resolver) sends an empty string to the server. A possible way would be `z.string().min(8).max(100).or(z.string().max(0))`, but it looks quite ugly ;-) It would be nice if `optional` would be supported in this case. That's why I opened an issue.

**Comment by medihack** [comment, marked as answer]
As more as I think about it, `z.string().min(8).max(100).or(z.string().max(0))` seems to be completely ok and `optional` is not the correct way to do this. I also closed the issue.

**Reply by scotttrinh** [reply]
Ahh, yeah, I thought you wanted optional in order to handle the `undefined` case, not the empty string. Another option is to use `z.literal("")` which _might_ perform better due to being a simpler parser (`===` vs. `.length`).

```ts
z.string().min(8).max(100).or(z.literal(""));
```

**Reply by medihack** [reply]
`literal` sounds great! Thanks!
